assets/tests/hooks/useDetour.test.ts

```diff
@@ -60,15 +60,20 @@ describe("useDetour", () => {
     expect(result.current.waypoints).toEqual([])
   })
 
-  test("when `addWaypoint` is called, `detourShape` is updated", async () => {
+  test("when `addWaypoint` is called, should update `detourShape` and `directions`", async () => {
     const start = { lat: 0, lon: 0 }
     const end = { lat: 1, lon: 1 }
 
-    jest
-      .mocked(fetchDetourDirections)
-      .mockImplementation((coordinates) =>
-        Promise.resolve({ coordinates: coordinates })
-      )
+    jest.mocked(fetchDetourDirections).mockImplementation((coordinates) =>
+      Promise.resolve({
+        coordinates: coordinates,
+        directions: [...Array(coordinates.length)].map((_, idx) => ({
+          instruction: idx.toString(),
+          name: idx.toString(),
+          type: "left",
+        })),
```

Comment:
```suggestion
        directions: [
          { instruction: "Turn left on Main St", name: "Main St", type: "left" },
          { instruction: "Slight right on Centre Ave", name: "Centre Ave", type: "slight_right" },
        ],
```

(Along with the corresponding change to the assertion)

I find tests easier to read when they have more explicit test data, and when two things are supposed to be exactly the same, like the `directions` return value from `fetchDetourDirections` and `directions` in the `result.current` assertion below, I'd prefer that exactly-the-sameness to be more explicit. Having these obviously-canned values in the setup and assertion does that.

(Using variables also works for that, like `start` and `end` above, but I tend to prefer the super explicit do-everything-as-literals approach)

Basically, I'd rather not have to spend time thinking about how `[...Array(coordinates.length)].map(...)` works unless that's a core part of what the test is asserting, which I don't think it is in this case.

Reply:
(posting for discussion)
We could use factories here to abstract away the actual generation of the data so we're not also having to handle that mental overhead _in_ the tests.

A factory for this could look like the following code block, and it could be improved by using the https://fakerjs.dev library.

```ts
import { Factory } from "fishery";
import { DetourShape, DirectionType } from "../../src/detour";
import { localGeoCoordinateFactory } from "./geoCoordinate";

export const directionTypeFactory = Factory.define<DirectionType>(({sequence}) => {
  const directions: DirectionType[] = [
    "depart",

    "sharp_right",
    "left",
    "keep_right",
    "sharp_left",
    "right",
    "keep_left",

    "straight",

    "enter_roundabout",
    "exit_roundabout",

    "u_turn",

    "goal",
  ]
  return directions[sequence % directions.length]
})

export const directionNameFactory = Factory.define<string>(({sequence}) => {
  const things = [
    "Main St",
    "High St",
    "Park Plaza",
    "Boylston St",
    "Massachusetts Ave",
  ]
  return things[sequence % things.length]
})

const makeInstruction = (type: DirectionType, name: string) => {
  switch (type) {
    case "left":
      return `Turn Left onto ${name}`
    case "right":
      return `Turn Right onto ${name}`
    case "keep_left":
      return `Keep Left on ${name}`
    case "keep_right":
      return `Keep Right on ${name}`
    case "sharp_left":
      return `Make a Sharp Left on ${name}`
    case "sharp_right":
      return `Make a Sharp Right on ${name}`
    case "slight_left":
      return `Make a Slight Left on ${name}`
    case "slight_right":
      return `Make a Slight Right on ${name}`
    case "enter_roundabout":
      return `Enter the roundabout and take the 1st exit onto ${name}`
    case "exit_roundabout":
      return `Exit the roundabout at the the 1st exit onto ${name}`
    case "straight":
      return `Continue Straight on ${name}`
    case "depart":
      return `Depart from ${name}`

    default:
      `${type} onto ${name}`;
  }
}

export const directionsFactory = Factory.define<DetourShape["directions"][0]>(({sequence}) => {
  const name = directionNameFactory.build();
  const type = directionTypeFactory.build();
  return ({
    instruction: makeInstruction(type, name),
    name,
    type,
  });
})

export const detourShapeFactory = Factory.define<DetourShape>(({}) => ({
  coordinates: localGeoCoordinateFactory.buildList(3).map(v => ({ lat: v.latitude, lon: v.longitude})),
  directions: directionsFactory.buildList(3)
}))
```

Usage of this factory would look like
```ts
const detourShape = detourShapeFactory.build()
```

I'm also not even certain that we should be mocking out instruction data in `makeInstruction`, as I think the code should treat it as an opaque blob.

Reply:
TL;DR...

- I definitely want to see us either explicitly use factories, or at least use objects that sort of behave like them. I think that'll really help to reduce boilerplate. 
- I strongly prefer to have any data that's asserted on in a test be because of something that was set up explicitly in that test, rather than relying on factory defaults.
- I weakly prefer factories to have their defaults be the simplest possible values that make them valid.
- I agree that we probably don't need `makeInstruction`, and in general I prefer to avoid having too much nontrivial logic in tests.

---

> We could use factories here to abstract away the actual generation of the data so we're not also having to handle that mental overhead in the tests.

If the test cares about a particular chunk of data, then IMO, it creates more overhead if the data generation happens somewhere else.

For instance, a test like:

```javascript
test("handles directions properly", () => {
  const apiData = detourShapeFactory.build()
  jest.mocked(fetchTheThings).mockReturnValue(apiData)

  const directionsContainer = however.we.find.that.on.the.page()

  expect(directionsContainer).toInclude("Depart from Main St")
})
```

In the context of that test, the `"Depart from Main St"` comes out of nowhere, and I have to figure out whether that data came from the factory, some default in the code, or somewhere else. If I saw a test like that, I would assume that it was saying that the _production_ code should default to `"Depart from Main St"` if it wasn't given explicit directions. 

Even something like this:

```javascript
test("handles directions properly", () => {
  const apiData = detourShapeFactory.build()
  jest.mocked(fetchTheThings).mockReturnValue(apiData)

  const directionsContainer = however.we.find.that.on.the.page()

  expect(directionsContainer).toInclude(extractAUsefulStringFrom(apiData.directions))
})
```

...still doesn't give us much insight into what `directions` actually _is_. I don't know if there are any edge cases being validated here, like maybe the factory defaults directions to `null`, or `[]`. (Not to mention that `extractAUsefulStringFrom()` is its own can of worms, and could wind up being almost a re-implementation of the production code)

This is what I want to see:

```javascript
test("handles directions properly", () => {
  const apiData = detourShapeFactory.build({
    directions: [{ instruction: "Turn Right on Centre Blvd" }, /* Etc */]
  })
  jest.mocked(fetchTheThings).mockReturnValue(apiData)

  const directionsContainer = however.we.find.that.on.the.page()

  expect(directionsContainer).toInclude("Turn Right on Centre Blvd")
})
```

So the data in the assertion is set up explicitly when we invoke the factory.

That said, I think factories have their place here very solidly. `apiData` needs to have both `directions` and `coordinates` in order to avoid causing errors, but I like that ☝️ test doesn't say anything about `coordinates`. It just lets the factory go ahead and put _something_ in `coordinates` so that the test can run, but then lets us focus on `directions`-y stuff. 

Basically, they save us from something like 👇 

```javascript
test("handles directions properly", () => {
  const apiData = createAnApiData({
    coordinates: [], // has to be here so that the test doesn't console.error, but we don't actually care
    directions: [{ instruction: "Turn Right on Centre Blvd" }, /* Etc */]
  })
  jest.mocked(fetchTheThings).mockReturnValue(apiData)

  const directionsContainer = however.we.find.that.on.the.page()

  expect(directionsContainer).toInclude("Turn Right on Centre Blvd")
})
```

Where the parameters to `createAnApiData()` get more and more annoying as the API gets more sophisticated.

---

Related to ☝️, I have a weak preference for factories to have the simplest defaults possible, so in this case, `coordinates` would default to `[]` unless we explicitly set it to something else. It's a weak preference, mostly related to the fact that if data in the test has anything interesting going on, then I'd rather have that be more obvious in the test setup itself.

---

> I'm also not even certain that we should be mocking out instruction data in `makeInstruction`, as I think the code should treat it as an opaque blob.

Yeah I agree that we don't need this. Having it would set us up for a "Wait, where did `"Make a Sharp Right on Main St"` come from??" situation, because I'm not sure that it would be obvious that it came from the factory rather than the production code.

In general, I prefer to avoid having nontrivial logic in tests. Enough to get us away from annoyingly repetitive boilerplate, sure, but I'd mostly rather see the data exactly as it's going into the system-under-test, and then assert on the data exactly as it comes back out.

Reply:
I'll caveat this by saying that I haven't read the whole PR or fully processed every bit of this discussion, but my $0.02 is that I think I agree with @joshlarson that if the specific values in a factory-generated piece of data matter for the test, the test should set those explicitly rather than relying on factory defaults..

Reply:
PTALA [`78f43f6` (#2367)](https://github.com/mbta/skate/pull/2367/commits/78f43f6b4f09e0b3e327acd95aa61e0587079b6b) 🙏🏻